### PR TITLE
Add `RecordReader` trait and proc macro to implement it for a struct

### LINKED ISSUE #4772: There exists a `ParquetRecordWriter` proc macro in `parquet_derive`, but `ParquetRecordReader` is missing
## Is your feature request related to a problem or challenge? Please describe what you are trying to do.
Reading parquet files into slices of structs can take quite a lot of code. There exists a `parquet_derive::ParquetRecordWriter` derive macro to write from a slice of structs to a parquet file, but there is no equivalent `parquet_derive::ParquetRecordReader` macro.

## Describe the solution you'd like
A `parquet_derive::ParquetRecordReader` macro that does the same as `parquet_derive::ParquetRecordWriter` but for reading.

There already exists a `parquet::record::RecordWriter` trait:
https://github.com/apache/arrow-rs/blob/587250c8e0f9707cc102bd04573395c153249ced/parquet/src/record/record_writer.rs#L23-L31

I would like there to be a similar `parquet::record::RecordReader` trait such as:
```rust
pub trait RecordReader<T> {
    fn read_from_row_group(
        &mut self,
        row_group_reader: &mut dyn RowGroupReader,
        max_records: usize,
      ) -> Result<(), ParquetError>;
}
```

There also exists a `parquet_derive::ParquetRecordWriter` proc macro to implement this trait for slices of structs:
https://github.com/apache/arrow-rs/blob/587250c8e0f9707cc102bd04573395c153249ced/parquet_derive/src/lib.rs#L77-L78
Generates code to implement trait here:
https://github.com/apache/arrow-rs/blob/587250c8e0f9707cc102bd04573395c153249ced/parquet_derive/src/lib.rs#L98

So I would again like there to be a similar `parquet_derive::ParquetRecordReader` that implements something like:
```rust
impl #generics ::parquet::record::RecordReader<#derived_for #generics> for &mut [#derived_for #generics] {
```

## Describe alternatives you've considered
The alternative is for the user to write this code by hand. However, for large structs a macro is necessary and then quickly becomes quite a lot of code ontop of the existing parquet library.

## Additional context
I have already implemented a possible solution which I will make a PR for.

### PR DESCRIPTION
**Pull request by Joseph-Rance**
# Which issue does this PR close?
Closes #4772.

# What changes are included in this PR?
 - Added a `parquet::record::RecordReader` trait similar to the existing `parquet::record::RecordWriter`
 - Added a `parquet_derive::ParquetRecordReader` derive macro to implement the new `RecordReader` trait for slices of structs, similar to the existing `parquet_derive::ParquetRecordWriter` macro
 - Added tests to `parquet_field.rs` for `reader_snipped` similar to the existing tests for `writer_snipped`
 - Added a test to `parqued_derive_test` to test writing a struct slice to a parquet file, reading it again, and then checking the output is the same as the original struct.
 - Not all types supported by `ParquetRecordWriter` are supported by `ParquetRecordReader` yet. Specifically there is no support for options or references. References is difficult because using `Rc` would mean we have to manually annotate types for the compiler which requires quite a lot of matching on the `Field`'s `ty` field. See: https://github.com/Joseph-Rance/arrow-rs/blob/f49472eb93c4bf27aa20d8040c57539fc73d1059/parquet_derive/src/parquet_field.rs#L452-L454

 - Updated the readme to reflect the above point

# Are there any user-facing changes?
 - Users will now have access to the `parquet::record::RecordReader` trait and the `parquet_derive::ParquetRecordReader` proc macro

### CODE REVIEW

**Comment by Joseph-Rance** [comment]
@tustvold would it be possible to get a review on this? Apologies if I am pinging the wrong person. Please do let me know if there is anything further I need to do to get this ready for review

**Comment by tustvold** [comment]
Apologies, it is on my list to review, but I'm not very familiar with the non-arrow APIs so need longer to sit down and get the necessary context. Development has largely focused over the last couple of years on the columnar, arrow-based APIs as they have better performance characteristics, so I lack any major context 😅

**Comment by Joseph-Rance** [comment]
No problem. Just wanted to make sure this PR hadn't been missed

**Comment by Joseph-Rance** [comment]
@tustvold Apologies for pinging you again. I was wondering if you could estimate a possible timeline for review of this PR?

File: parquet_derive/src/parquet_field.rs [hunk 1]
```diff
@@ -354,6 +428,50 @@ impl Field {
...
 
+    fn copied_direct_fields(&self) -> proc_macro2::TokenStream {
+        let field_name = &self.ident;
+        let is_a_byte_buf = self.is_a_byte_buf;
+        let is_a_timestamp =
+            self.third_party_type == Some(ThirdPartyType::ChronoNaiveDateTime);
+        let is_a_date = self.third_party_type == Some(ThirdPartyType::ChronoNaiveDate);
+        let is_a_uuid = self.third_party_type == Some(ThirdPartyType::Uuid);
+
+        let value = if is_a_timestamp {
```
Comment (tustvold) [inline parquet_derive/src/parquet_field.rs:439]: This might be more clearly written as a match block

File: parquet_derive/src/parquet_field.rs [hunk 2]
```diff
@@ -354,6 +428,50 @@ impl Field {
...
+            quote! { ::chrono::naive::NaiveDateTime::from_timestamp_millis(vals[i]).unwrap() }
+        } else if is_a_date {
+            quote! { ::chrono::naive::NaiveDate::from_num_days_from_ce_opt(vals[i]
```
Comment (tustvold) [inline parquet_derive/src/parquet_field.rs:442]: I think this would be better expressed using signed_duration_since

File: parquet_derive/src/parquet_field.rs [hunk 3]
```diff
@@ -354,6 +428,50 @@ impl Field {
...
+            - ::chrono::naive::NaiveDate::from_ymd_opt(0, 12, 31).unwrap()).num_days()) as i32).unwrap() }
+        } else if is_a_uuid {
+            quote! { ::uuid::Uuid::parse_str(vals[i].data().convert()).unwrap() }
+        } else if is_a_byte_buf {
+            quote! { vals[i].data().convert() }
+        } else {
+            quote! { vals[i].convert() }
+        };
+
+        // The below code would support references in field types, but it prevents the compiler
```
Comment (tustvold) [inline parquet_derive/src/parquet_field.rs:453]: How would references work, who would own the allocation from which the field borrows?

File: parquet/src/record/record_reader.rs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use super::super::errors::ParquetError;
+use super::super::file::reader::RowGroupReader;
+
+pub trait RecordReader<T> {
```
Comment (tustvold) [inline parquet/src/record/record_reader.rs:21]: Perhaps a doc comment, in particular explaining what T is

**Review by tustvold — COMMENTED** [review]
I've done an initial review, overall this looks very nice, I would still encourage people who care about performance to use the columnar APIs, but I can see the appeal of a row-API.

Overall this looks well tested and seems to make sense. My only comment is that the logic that converts from columnar to the row representation I think could be made easier to follow and a bit better documented

File: parquet_derive/src/parquet_field.rs [hunk 5]
```diff
@@ -682,6 +838,66 @@ impl Type {
     }
 }
 
+// returns quote of function needed to convert from parquet
+// types back into the types used in the struct fields.
+pub fn get_convertable_quote() -> proc_macro2::TokenStream {
```
[the remark this answers is not in the document]
Comment (Joseph-Rance) [reply parquet_derive/src/parquet_field.rs:843]: To add to this, I'm not actually sure what the compile/runtime cost of this would be. I think runtime would be the same, but at compile time I'm not sure the difference. I can't imagine there will be much change, though.

[the remark this answers is not in the document]
Comment (tustvold) [reply parquet_derive/src/parquet_field.rs:843]: I'm not sure the reader as formulated can handle references nor options, so should be just a case of matching the concrete types the trait is defined for.

In the interests of making the code easier to maintain in future, I think this would be a good idea

File: parquet_derive/src/parquet_field.rs (hunk 1 above)
Comment (Joseph-Rance) [reply parquet_derive/src/parquet_field.rs:439]: I did this to mirror what already exists in `copied_direct_vals`, and I agree match would be nicer so I'll update both

File: parquet_derive/src/parquet_field.rs (hunk 3 above)
Comment (Joseph-Rance) [reply parquet_derive/src/parquet_field.rs:453]: I think because I used an `Rc` type here, we don't need to have something owning the value referenced. It's possible I have misunderstood how this works, though. [The documentation](https://doc.rust-lang.org/alloc/rc/#:~:text=%E2%80%98Rc%E2%80%99%20stands%20for%20%E2%80%98Reference%20Counted%E2%80%99.%20The%20type%20Rc%3CT%3E,pointer%20to%20the%20same%20allocation%20in%20the%20heap.) says:

> The type [Rc<T>](https://doc.rust-lang.org/alloc/rc/struct.Rc.html) provides shared ownership of a value of type T, allocated in the heap. Invoking [clone](https://doc.rust-lang.org/core/clone/trait.Clone.html#tymethod.clone) on [Rc](https://doc.rust-lang.org/alloc/rc/struct.Rc.html) produces a new pointer to the same allocation in the heap. When the last [Rc](https://doc.rust-lang.org/alloc/rc/struct.Rc.html) pointer to a given allocation is destroyed, the value stored in that allocation (often referred to as “inner value”) is also dropped.

Comment (tustvold) [reply parquet_derive/src/parquet_field.rs:453]: If you have a reference such as `&'a str` wrapping it up in a, e.g. `Rc<&'a str>` doesn't remove the lifetime. To remove the lifetime you would need to create a `Rc<str>` which will perform a new allocation and clone the string data. In this case using `String` would be equivalent

File: parquet/src/record/record_reader.rs (hunk 4 above)
Comment (Joseph-Rance) [reply parquet/src/record/record_reader.rs:21]: I added the `T` because it was included in `RecordWriter`, but removing it gives the following error:
```
error[E0117]: only traits defined in the current crate can be implemented for arbitrary types
  --> parquet_derive_test/src/lib.rs:52:42
   |
52 | #[derive(PartialEq, ParquetRecordWriter, ParquetRecordReader, Debug)]
   |                                          ^^^^^^^^^^^^^^^^^^^
   |                                          |
   |                                          impl doesn't use only types from inside the current crate
   |                                          this is not defined in the current crate because slices are always foreign
   |
   = note: define and implement a trait or new type instead
   = note: this error originates in the derive macro `ParquetRecordReader` (in Nightly builds, run with -Z macro-backtrace for more info)
```
Where the generated code with the generic included is:
```
impl ::parquet::record::RecordReader<APartiallyCompleteRecord> for &mut [APartiallyCompleteRecord] 
```
I don't entirely understand why the type parameter solves this issue, but I do get the feeling it is more specific to the implementation than the trait itself. I would struggle to explain why `T` is necessary without referring directly to this macro. I'm honesty a little lost here. Do you think there is a better solution / do you know why the type parameter actually solves the error?

File: parquet_derive/src/parquet_field.rs (hunk 2 above)
Comment (Joseph-Rance) [reply parquet_derive/src/parquet_field.rs]: fixed in [9f2dd9f](https://github.com/apache/arrow-rs/pull/4773/commits/9f2dd9f901462b6221bc7ba9fab04343c85469b6)

File: parquet_derive/src/parquet_field.rs (hunk 1 above)
Comment (Joseph-Rance) [reply parquet_derive/src/parquet_field.rs]: fixed in [9f2dd9f](https://github.com/apache/arrow-rs/pull/4773/commits/9f2dd9f901462b6221bc7ba9fab04343c85469b6)

File: parquet_derive/src/parquet_field.rs [hunk 6]
```diff
@@ -223,6 +223,80 @@ impl Field {
...
+            vals_vec.resize(max_records, #default_type);
+            let mut vals: &mut [#parquet_type] = vals_vec.as_mut_slice();
+            if let #column_reader(mut typed) = column_reader {
+                typed.read_records(max_records, None, None, vals)?;
+            } else {
+                panic!("Schema and struct disagree on type for {}", stringify!{#ident});
+            }
+        };
+
+        let vals_writer = match &self.ty {
```
[the remark this answers is not in the document]
Comment (Joseph-Rance) [reply parquet_derive/src/parquet_field.rs:266]: fixed in [9f2dd9f](https://github.com/apache/arrow-rs/pull/4773/commits/9f2dd9f901462b6221bc7ba9fab04343c85469b6)

File: parquet_derive/src/parquet_field.rs [hunk 7]
```diff
@@ -223,6 +223,80 @@ impl Field {
...
+    ///
+    /// `Option` types and references not supported
+    pub fn reader_snippet(&self) -> proc_macro2::TokenStream {
+        use parquet::basic::Type as BasicType;
+
+        let ident = &self.ident;
+        let column_reader = self.ty.column_reader();
+        let parquet_type = self.ty.physical_type_as_rust();
+
+        let default_type = match self.ty.physical_type() {
```
[the remark this answers is not in the document]
Comment (Joseph-Rance) [reply parquet_derive/src/parquet_field.rs]: fixed in [9f2dd9f](https://github.com/apache/arrow-rs/pull/4773/commits/9f2dd9f901462b6221bc7ba9fab04343c85469b6)

File: parquet_derive/src/parquet_field.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Joseph-Rance) [reply parquet_derive/src/parquet_field.rs]: fixed in [9f2dd9f](https://github.com/apache/arrow-rs/pull/4773/commits/9f2dd9f901462b6221bc7ba9fab04343c85469b6)

File: parquet_derive/src/parquet_field.rs [hunk 8]
```diff
@@ -524,6 +665,21 @@ impl Type {
...
+
+        match self.physical_type() {
+            BasicType::BOOLEAN => quote! { bool },
+            BasicType::INT32 => quote! { i32 },
+            BasicType::INT64 => quote! { i64 },
+            BasicType::INT96 => unimplemented!("96-bit int currently is not supported"),
+            BasicType::FLOAT => quote! { f32 },
+            BasicType::DOUBLE => quote! { f64 },
+            BasicType::BYTE_ARRAY => quote! { ::parquet::data_type::ByteArray },
+            BasicType::FIXED_LEN_BYTE_ARRAY => quote! { ::parquet::data_type::ByteArray },
```
[the remark this answers is not in the document]
Comment (Joseph-Rance) [reply parquet_derive/src/parquet_field.rs]: fixed in [9f2dd9f](https://github.com/apache/arrow-rs/pull/4773/commits/9f2dd9f901462b6221bc7ba9fab04343c85469b6)

File: parquet_derive/src/parquet_field.rs (hunk 3 above)
Comment (Joseph-Rance) [reply parquet_derive/src/parquet_field.rs]: I wrote the below code to get the `Rc<str>`, but then I still can't get a normal reference from this. I guess that's what you were saying here?
```rust
Type::Reference(_, ref first_type) => match **first_type {
  Type::TypePath(_) => match self.ty.last_part().as_str() {
      "String" => quote! {
          std::rc::Rc::<String>::new(String::from(std::str::from_utf8(vals[i].data()).expect("invalid UTF-8 sequence")))
      },
      "str" => quote! {
          std::rc::Rc::<str>::from(std::str::from_utf8(vals[i].data()).expect("invalid UTF-8 sequence"))
      },
      f => unimplemented!("Unsupported: {:#?}", f),
  },
  Type::Slice(_) => quote! { std::rc::Rc::<str>::from(vals[i].data()) },
  ref f => unimplemented!("Unsupported: {:#?}", f),
},
```
In that case I guess we can't support returning references

**Comment by Joseph-Rance** [comment]
@tustvold I've resolved most of your comments. Left two open:
 - For the references, I think there isn't anything more to do, but I left it open in case I have misunderstood the situation
 - For the other comment, I wonder if you have any opinions on what I should write?

File: parquet/src/record/record_reader.rs (hunk 4 above)
Comment (tustvold) [reply parquet/src/record/record_reader.rs:22]: I think if instead of deriving the trait for a mutable slice you derived it for `Vec` it would remove the need for this, and would make for a cleaner API

File: parquet_derive/README.md [hunk 9]
```diff
@@ -77,16 +77,55 @@ writer.close_row_group(row_group).unwrap();
...
+    pub now: chrono::NaiveDateTime,
+    pub byte_vec: Vec<u8>,
+}
+
+// Initialize your parquet file
+let reader = SerializedFileReader::new(file).unwrap();
+let mut row_group = reader.get_row_group(0).unwrap();
+
+// create your records to read into
+let mut chunks = vec![ACompleteRecord{ ... }];
```
Comment (tustvold) [inline parquet_derive/README.md:105]: See comment above, I think if the trait was instead derived for Vec it would make a simpler interface and would remove the need for this default initialisation

File: parquet/src/record/record_reader.rs (hunk 4 above)
Comment (Joseph-Rance) [reply parquet/src/record/record_reader.rs:22]: Unfortunately, it looks like switching to `Vec` doesn't solve this:
```
22 | #[derive(ParquetRecordWriter)]
   |          ^^^^^^^^^^^^^^^^^^^
   |          |
   |          impl doesn't use only types from inside the current crate
   |          `Vec` is not defined in the current crate
```
I'm also not sure what the policy is here on changing existing functionality? I feel like we would want both writer and reader to be implemented for the same thing but changing the writer to `Vec` would break existing code that uses this.

Comment (tustvold) [reply parquet/src/record/record_reader.rs:22]: Aah yeah, because this is a macro used in an external crate to the trait definition, I think we should still derive for Vec instead of a mutable slice, as it makes the API significantly easier to use

Comment (Joseph-Rance) [reply parquet/src/record/record_reader.rs:22]: Ok. I'll change to `Vec`, but if we are keeping the `T` I'm not really sure how to explain why it is necessary in the trait definition?

File: parquet_derive/README.md (hunk 9 above)
Comment (Joseph-Rance) [reply parquet_derive/README.md:105]: Where do you want to initialise the structs? If we want to do it at the start of the macro we would need the struct to derive default, which I think might be problematic for some structs. I don't think we can very easily initialise the struct with all the data we have already read because we read them column by column. I can save the columns and then build the structs after the entire file is read but this seems possibly overcomplicated compared to just letting the user initialise the structs.

Comment (tustvold) [reply parquet_derive/README.md:105]: I must be missing something, why do you need to initialize the struct at all? Can't you just construct it from the parsed values? I might be missing something

File: parquet/src/record/record_reader.rs (hunk 4 above)
Comment (tustvold) [reply parquet/src/record/record_reader.rs:22]: I guess you just say it is work around the orphan rules, it is a bit gross and I don't like it but that appears to be what the writer does so... :shrug:

File: parquet_derive/README.md (hunk 9 above)
Comment (Joseph-Rance) [reply parquet_derive/README.md:105]: We could, but currently we iterate over the columns of the parquet file, so if we have a struct with fields `a` and `b`, then we would read and set all `a` values, and then all `b` values. However, to construct the struct directly we need `a` and `b` at the same time. So to do this we would have to read the entire file into slices of each field and then after that we could construct the structs.

I guess there is no problem with this other than that it wastes some memory by holding the entire file in memory twice (once for the structs we create and once (temporarily) for the values we are putting in the structs) to save one line of code.

Comment (tustvold) [reply parquet_derive/README.md:105]: Aren't you already doing this though? read_records returns a slice of values for a particular column?

Comment (Joseph-Rance) [reply parquet_derive/README.md:105]: Yes but we are immediately writing that into the struct so we aren't necessarily holding all of these slices in memory at once. To be honest, such a minor performance consideration isn't realistically going to be an issue.

Comment (Joseph-Rance) [reply parquet_derive/README.md:105]: I spent some time trying to get this to work, but it requires quite a lot of implementation (or I am doing something wrong) because we need to somehow generate identifiers for each slice and then feed them through to be used later when constructing the structs.

It is definitely possible to do this, but I think I have found a slightly easier way to do this. What do you think of the changes I just pushed? All fields we currently support implement default so I think this should have identical behaviour for the user's point of view.

**Comment by Joseph-Rance** [comment]
Any idea why clippy is suddenly failing on code that (I think) has nothing to do with this PR? Can I ignore these issues or is it due to something I have changed?

**Comment by tustvold** [comment]
You probably need to merge the latest master to pick up fixes for the lints added in the latest Rust version

**Comment by Joseph-Rance** [comment]
@tustvold could you approve the workflows please?

**Comment by Joseph-Rance** [comment]
Workflows should pass now

File: parquet_derive/src/parquet_field.rs [hunk 10]
```diff
@@ -348,6 +417,47 @@ impl Field {
...
+    fn copied_direct_fields(&self) -> proc_macro2::TokenStream {
+        let field_name = &self.ident;
+
+        let value = match self.third_party_type {
+            Some(ThirdPartyType::ChronoNaiveDateTime) => {
+                quote! { ::chrono::naive::NaiveDateTime::from_timestamp_millis(vals[i]).unwrap() }
+            }
+            Some(ThirdPartyType::ChronoNaiveDate) => {
+                quote! {
+                    ::chrono::naive::NaiveDate::from_num_days_from_ce_opt(vals[i]
```
Comment (tustvold) [inline parquet_derive/src/parquet_field.rs:429]: Why is this logic different to above?

File: parquet_derive/src/lib.rs [hunk 11]
```diff
@@ -95,7 +95,7 @@ pub fn parquet_record_writer(input: proc_macro::TokenStream) -> proc_macro::Toke
         field_infos.iter().map(|x| x.parquet_type()).collect();
 
     (quote! {
-    impl #generics ::parquet::record::RecordWriter<#derived_for #generics> for &[#derived_for #generics] {
+    impl #generics ::parquet::record::RecordWriter<#derived_for #generics> for Vec<#derived_for #generics> {
```
Comment (tustvold) [inline parquet_derive/src/lib.rs:98]: Why this change? I think it makes sense that the writer doesn't require an owned container as it isn't mutating it?

File: parquet_derive/src/lib.rs [hunk 12]
```diff
@@ -137,3 +137,89 @@ pub fn parquet_record_writer(input: proc_macro::TokenStream) -> proc_macro::Toke
...
+/// use parquet::file::{serialized_reader::SerializedFileReader, reader::FileReader};
+/// use parquet_derive::{ParquetRecordReader};
+///
+/// #[derive(ParquetRecordReader)]
+/// struct ACompleteRecord {
+///     pub a_bool: bool,
+///     pub a_string: String,
+/// }
+///
+/// pub fn read_some_records() -> ACompleteRecord {
```
Comment (tustvold) [inline parquet_derive/src/lib.rs:161]:
```suggestion
/// pub fn read_some_records() -> Vec<ACompleteRecord> {
```
Perhaps?

File: parquet_derive/README.md [hunk 13]
```diff
@@ -71,22 +71,61 @@ let mut row_group = writer.next_row_group().unwrap();
...
+- [x] Derive definition_levels for `Option` for writing
+- [ ] Derive definition levels for nested structures for writing
 - [ ] Derive writing tuple struct
 - [ ] Derive writing `tuple` container types
 
+- [x] Support reading `String`, `&str`, `bool`, `i32`, `f32`, `f64`, `Vec<u8>`
+- [ ] Support reading/writing dictionaries
+- [x] Support reading/writing logical types like timestamp
+- [ ] Derive definition_levels for `Option` for reading
+- [ ] Derive definition levels for nested structures for reading
```
Comment (tustvold) [inline parquet_derive/README.md:125]:
```suggestion
- [ ] Handle definition_levels for `Option` for reading
- [ ] Handle definition levels for nested structures for reading
```

**Review by tustvold — APPROVED** [review]
This looks reasonable to me, I have to confess to not being sure that it will be possible to evolve this interface to support nested data, but perhaps that isn't a problem

File: parquet_derive/src/parquet_field.rs (hunk 10 above)
Comment (Joseph-Rance) [reply parquet_derive/src/parquet_field.rs:429]: I'm not sure exactly what your question related to, but this logic for getting the `NaiveDate` is just the inverse of above. However, there is no function `from_signed_duration_since` that mirrors the `signed_duration_since` function we are using above, so I had to use `from_num_days_from_ce_opt`, which uses a different date to measure days from. That's why we need to add on the number of days between that other date and 01/01/1970.

File: parquet_derive/src/lib.rs (hunk 11 above)
Comment (Joseph-Rance) [reply parquet_derive/src/lib.rs:98]: I made the change to make the reader and writer both be for the same type. On second thought perhaps also not worth breaking existing code that uses this macro for no reason, so I'll switch back.

File: parquet_derive/README.md (hunk 13 above)
Comment (Joseph-Rance) [reply parquet_derive/README.md:125]: I assume I should also make this change for lines 116 and 117?

Comment (tustvold) [reply parquet_derive/README.md:125]: Lines 116 and 117 are correct, the writer derives levels from the input data, the reader, however, needs to do the opposite :smile:

File: parquet_derive/src/parquet_field.rs [hunk 14]
```diff
@@ -348,6 +418,48 @@ impl Field {
...
+                + ((::chrono::naive::NaiveDate::from_ymd_opt(1970, 1, 1).unwrap()
+                        .signed_duration_since(
+                            ::chrono::naive::NaiveDate::from_ymd_opt(0, 12, 31).unwrap()
```
Comment (tustvold) [inline parquet_derive/src/parquet_field.rs:434]:
```suggestion
                            ::chrono::naive::NaiveDate::from_num_days_from_ce_opt(0)
```
I found the use of 12, 31 rather confusing
